pkg/controller/job/success_policy.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"fmt"
+
+	batch "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+func canApplySuccessPolicy(jobCtx *syncJobCtx) bool {
+	return feature.DefaultFeatureGate.Enabled(features.JobSuccessPolicy) && jobCtx.finishedCondition == nil &&
+		jobCtx.job.Spec.SuccessPolicy != nil && len(jobCtx.job.Spec.SuccessPolicy.Criteria) != 0
+}
+
+func metSuccessPolicy(logger klog.Logger, successPolicy *batch.SuccessPolicy, completions int32, succeededIndexes orderedIntervals) (string, bool) {
+	if successPolicy == nil {
+		return "", false
+	}
+
+	criteriaMetMsg := "Met criteria at index"
+	for index, criteria := range successPolicy.Criteria {
+		if criteria.SucceededIndexes != nil {
+			requiredIndexes := parseIndexesFromString(logger, *criteria.SucceededIndexes, int(completions))
+			// Failed to parse succeededIndexes due to some errors like invalid format.
+			if len(requiredIndexes) == 0 {
+				continue
+			}
+			diff, met := requiredIndexes.isProperSubsetOf(succeededIndexes)
```

Comment:
Why `proper` subset? If the sets are equal the criteria are met too. Maybe `isSubset` or `isContainedIn`.

Reply:
> If the sets are equal the criteria are met too

That makes sense. This was brought by my misunderstanding of "proper subset".
We should use the `isSubset`. Thanks.

Reply:
Done.